service/service.go

```diff
@@ -686,12 +687,65 @@ func (s *Service) MakeNewSecret(ctx context.Context, req *pbsvc.UserRequest) (*p
 }
 
 // VerifyEmailToken checks if received token is found in the email_tokens table.
-// If found and token is NOT expired, returns OK.
-// If found, but token IS expired, the token row with expired token will be deleted,
-// service will generate a new email token and insert into email_tokens table,
-// and return ERROR with token expired message.
-// If token is not found, return ERROR with token does not exist message.
+// If found and token is NOT expired, deletes token row and returns OK.
+// If found, but token IS expired, it will return a expired token error.
+// Additionally for expired tokens, if user is new, it will delete token AND user row, else just deletes the token row.
+// If token is not found, return error with token does not exist message.
 func (s *Service) VerifyEmailToken(ctx context.Context, req *pbsvc.UserRequest) (*pbsvc.UserResponse, error) {
-	// TODO
-	return nil, nil
+	logger.RequestService("VerifyEmailToken")
+
+	if ok := serviceStateLocker.isStateAvailable(); !ok {
+		logger.Error(consts.VerifyEmailToken, consts.ErrServiceUnavailable.Error())
+		return nil, consts.ErrStatusServiceUnavailable
+	}
+
+	if req == nil {
+		return nil, consts.ErrStatusNilRequestUser
+	}
+
+	if err := refreshDBConnection(); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	emailToken := req.GetIdentification().GetToken()
```

Comment:
`Identification` can be nil

Reply:
I always wondered about that. I don't know if grpc does something to these objects or it is Golang specific thing, but req object isn't like javascript where it throws an error when trying to access a property on a null object. Checking for req like below i found is not necessary either
```
if req == nil {
		return nil, consts.ErrStatusNilRequestUser
	}
```
Because if req is nil, everything inside it is nil, so it returns nil even if you try to access properties.

With that said, `req.GetIdentification().GetToken()` will return nil if `Identification` is nil, I already unit tested it. Do you still want me to check it?

Reply:
In Java, this is `NullPointerException`
In C++, this is `SegmentationFault`
I am sure other languages will have their own issue if `Identification` is nil. `Identification` here by the is a pointer type. I am trying to find the spec about it, but this is still a bad practice.

To give you a better idea check this out https://tour.golang.org/methods/12